src/wallet/db.cpp

```diff
@@ -570,6 +571,16 @@ void BerkeleyEnvironment::CloseDb(const std::string& strFile)
 
 void BerkeleyEnvironment::ReloadDbEnv()
 {
+    // Make sure that no Db's are in use
+    AssertLockNotHeld(cs_db);
+    std::unique_lock<std::recursive_mutex> lock(cs_db);
```

Comment:
It would be slightly cleaner to use `std::unique_lock<CCriticalSection>` here.

Also, please squash this into the first commit - currently your PR introduces a bug and then fixes it later.

We should probably also work on making the sync.h primitives more C++11 compatible, but let's leave that for later.

Reply:
Done